### PR TITLE
Refactor Role and add some user management API

### PR DESCRIPTION
**Pull request by jonmv**
@tokle you can review here, but I'll need to make a new PR against master when https://github.com/vespa-engine/vespa/pull/9002 is merged.

### CODE REVIEW

**Comment by jonmv** [comment]
@mpolden feel free to have an opinion on the first commit, as well. I basically renamed the original `Role` to `ProtoRole`, and replaced the `RoleMembership` with a bound `Role`, which has a `Context` in it. Each type of `Context` has a different sub-class, and these can only be instantiated (outside the package) by a `Roles`, which is created with a given `SystemName`, and produces `Role`s only inside that. 

This solves several problems:
1. The correct `Context` for each `ProtoRole` is always ensured. 
2. The role definitions in `ProtoRole` are again an enum, allowing `switch` and easier reading. 
3. The `SystemName` part of the `Context` becomes basiaclly invisible to the user.

**Comment by jonmv** [comment]
The second commit introduces an API for use management. I envision the `CloudAccessControl` will be able to use only this interface to do its access control setup, and likewise for the user management API to come. 
I also believe this interface (`UserManagement`) is general enough to allow for different backends. 
That is, it should be possible to write all the `controller-server` code backend agnostic.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/GroupId.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import com.yahoo.vespa.hosted.controller.api.role.TenantRole;
+
+import java.util.Objects;
+
+/**
+ * An identifier for a group of {@link UserId}s, corresponding to a bound {@link Role}.
+ *
+ * @author jonmv
+ */
+public class GroupId {
```
Comment (tokle) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/GroupId.java]: I don't understand why you're introducing group here. We're creating/deleting/updating roles, not groups. The fact that some backends might choose to use groups for implementing the functionality shouldn't leak into the api.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/UserManagement.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Management of {@link UserId}s and {@link GroupId}s, used for access control with {@link Role}s.
+ *
+ * @author jonmv
+ */
+public interface UserManagement {
+
+    /** Creates the given group, or throws if the group already exists. */
+    void createGroup(GroupId group);
```
Comment (tokle) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/UserManagement.java]: -> createRole

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/UserManagement.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    /** Deletes the given group, or throws if it doesn't already exist.. */
+    void deleteGroup(GroupId group);
```
Comment (tokle) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/UserManagement.java]: deleteRole

**Review by tokle — CHANGES_REQUESTED** [review]
Good that we can finally hide the systemname! 

First commit lgtm, had some comments on the 2nd.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
+package com.yahoo.vespa.hosted.controller.api.role;
+
+import java.util.EnumSet;
+import java.util.Set;
+
+/**
+ * This declares all tenant roles known to the controller. A role contains one or more {@link Policy}s which decide
+ * what actions a member of a role can perform.
+ *
+ * Optionally, some role definition also inherit all policies from a "lower ranking" role. Read the list of roles
```
Comment (mpolden) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java]: definition -> definitions

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ * from {@code everyone} to {@code tenantAdmin}, in order, to see what policies these roles.
```
Comment (mpolden) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java]: these roles -> these roles have(?)

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * See {@link Role} for roles bound to a context, where policies can be evaluated.
+ *
+ * @author mpolden
+ * @author jonmv
+ */
+public enum ProtoRole {
```
Comment (mpolden) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java]: How about `RoleDefinition`? That what's you call it in the doc anyway. 🙂

**Review by mpolden — COMMENTED** [review]
Great!

(only reviewed 1st commit)

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/GroupId.java (hunk 1 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/GroupId.java]: As I said, no strong opinion on the name. `RoleId` sounds good?

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java (hunk 6 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java]: Indeed.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java (hunk 4 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java]: Yes.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java (hunk 5 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/ProtoRole.java]: Thanks.

**Comment by jonmv** [comment]
Reopening against master.

**Pull request closed without merging**